src/TS3Connection.ts

```diff
@@ -440,34 +445,25 @@ export class TS3Listener extends events.EventEmitter {
      * - a mapping of the TS-UID to the Discord-username is created
      */
     private async tagUp(g: discord.Guild, commander: Commander) {
-        let displayname = commander.getTS3DisplayName();
-        log("info", `Tagging up ${displayname} in ${g.name}.`);
+        log("info", `Tagging up ${(commander.getTS3DisplayName())} in ${g.name}.`);
         const registration = this.botgartClient.registrationRepository.getUserByAccountName(commander.getAccountName());
-        if (registration) {
+
+        let duser: discord.GuildMember | undefined;
+        if (registration !== undefined) {
             // the commander is member of the current discord -> give role
-            const crole = g.roles.cache.find(r => r.name === this.commanderRole);
-            const duser: discord.GuildMember | undefined = await g.members.fetch(registration.user); // cache.find(m => m.id === registration.user);
+            duser = await g.members.fetch(registration.user); // cache.find(m => m.id === registration.user);
             if (duser === undefined) {
                 log("warning", `Tried to find GuildMember for user with registration ID ${registration.user}, but could not find any. Maybe this is a caching problem?`)
             }
             commander.setDiscordMember(duser);
-            if (crole && commander.getDiscordMember()) {
-                await commander.getDiscordMember()?.roles.add(crole);
-            }
-            displayname = `${displayname} (${registration.registration_role})`;
-        }
-
-        // broadcast the message                    
-        const dchan: discord.TextChannel = <discord.TextChannel>g.channels.cache.find(c => c.name === this.broadcastChannel && c instanceof discord.TextChannel);
-        if (!dchan) {
-            log("warning", `I was supposed to broadcast the commander message on guild '${g.name}' in channel '${this.broadcastChannel}', but no such channel was found there. Skipping.`);
-        } else {
-            const pingRole = g.roles.cache.find(r => r.name === this.pingRole);
-            const channelPath = commander.getTs3channelPath().map(value => `\`${value}\``).join(" ❯ ");
 
-            const mes: string = L.get("COMMANDER_TAG_UP", [displayname, channelPath, pingRole ? pingRole.toString() : ""]);
-            dchan.send(mes);
+            await this.tagUpAssignRole(g, commander);
         }
+
+        await this.botgartClient.tagBroadcastService.sendTagUpBroadcast(g, commander, duser, registration)
```

Comment:
I am not sure why you are mixing `await` and callbacks here. If I am not mistaken, this should be:

```javascript
const message = await this.botgartClient.tagBroadcastService.sendTagUpBroadcast(g, commander, duser, registration);
commander.setBroadcastMessage(message)
```